### ISSUE TITLE
Non-generic subclass of generic class and narrowing

### ISSUE THREAD

**Issue by Hackrrr** [microsoft/pyright#4476]
**Describe the bug**
Suppose we have a generic class `A[T]` and we create a non-generic subclass `B`, e.g. `class B(A[int]): ...`.
Now consider several generic types, one of which is `A[T]`. If we narrow this one to `B`, pyright keeps all other occurrences of `T` same, although they could be narrowed (from `T` to `int` (assuming type correctness everywhere else)), but this current pyright behaviour can have some unwanted effects.

I'm not really sure if this issue is bug or feature request but it would be nice if pyright could handle such situations.

**Screenshots or Code**
There is simplified version of my "use case" that fails to type check:
```python
from typing import TypeVar, Generic, cast
T = TypeVar("T")

class DataProcessor(Generic[T]):
    def process_data(self, data: T):
        ...
class IntDataProcessor(DataProcessor[int]):
    def process_data(self, data: int):
        ...

def data_handler(value: T, processsor: DataProcessor[T]):
    if isinstance(processsor, IntDataProcessor):
        # error: Argument of type "T@data_handler" cannot be assigned to parameter "data" of type "int" in function "process_data"
        #   "object*" is incompatible with "int"
        processsor.process_data(value)
    else:
        processsor.process_data(value)  # all good here

    processsor.process_data(value)  # error again (same as before)

    processsor = cast(DataProcessor, processsor)  # ugly fix to remove narrowed `IntDataProcessor`
    processsor.process_data(value)  # ok now
```

**Expected behavior**
No error, everything will pass to type check.

**VS Code extension or command-line**
Command line; version 1.1.289

**Comment by erictraut** [comment]
I don't consider this a bug. Pyright is working as designed here.

With the `isinstance` type guard check, you're narrowing the type of `processor` to `IntDataProcessor`. The call `processor.process_data(value)` therefore should produce a type violation because `value` is type `T`, and `T` isn't type compatible with `int`.

What you're doing here is quite unusual and somewhat of an antipattern because your generic function is not acting in a truly generic manner; it's looking for specific subclasses of a generic type. I recommend looking for alternative approaches if you want this to work with static type checking. Mypy also generates an error for this code.

If you decide not to change your code, a cast or a `# type: ignore` are options for working around the issue.

**Comment by Hackrrr** [comment]
While I understand why pyright think that it's wrong, I would like you to (re)consider following snippet (which is essentially the original one reduced to the bare minimum):
```py
def simplified(value: T, processsor: DataProcessor[T]):
    if isinstance(processsor, IntDataProcessor):
        ...
    processsor.process_data(value)
```
Although this snippet is still kind of that antipattern, I think pyright should allow atleast this one (this case passes with mypy). But if nothing changes I actually understand the reasoning behind why such change would be rejected.
